### PR TITLE
dont block release on perf go.mod files

### LINKED ISSUE #17441: [Eng] Checks should not run on `testdata` directories
@chlowell was blocked on releasing azcore because the testdata/perf directory referenced the local version of internal/perf. The testdata directory is ignored by the go tooling, so we can ignore it as well.

### PR DESCRIPTION
**Pull request by seankane-msft**
@benbp based on my searches this was the most elegant way to tackle this situation. I don't want to block a release if a perf test's go.mod file uses a replace directive, this checks if `testdata` is in the path for a go.mod file and skips it if this is correct.

Closes #17441

### CODE REVIEW

File: eng/scripts/validate_go_mod.ps1 [hunk 1]
```diff
@@ -7,17 +7,19 @@ Push-Location sdk/$serviceDir
...
 }
 
 $hasError = $false
 foreach ($goMod in $goModFiles) {
-    $name = $goMod.FullName
-    $patternMatches = Get-Content $name | Select-String -Pattern "replace "
-    if ($patternMatches.Length -ne 0) {
-        Write-Host "Found a replace directive in go.mod file at $name"
-        $hasError = $true
+    if (-Not ($goMod.FullName -Like "*testdata*")) {
```
Comment (weshaggard) [inline eng/scripts/validate_go_mod.ps1:16]: I'd suggest using the common https://github.com/Azure/azure-sdk-for-go/blob/main/eng/scripts/Language-Settings.ps1#L37 they already return null for these non-standard go.mod files.

File: eng/scripts/Language-Settings.ps1 [hunk 2]
```diff
@@ -37,7 +37,7 @@ function Get-GoModuleVersionInfo($modPath)
 function Get-GoModuleProperties($goModPath)
 {
   $goModPath = $goModPath -replace "\\", "/"
-  if ($goModPath -match "(?<modPath>sdk/(?<serviceDir>(resourcemanager/)?([^/]+/)?(?<modName>[^/]+$)))")
+  if ($goModPath -match "(?<modPath>sdk/(?<serviceDir>(\w*/)?([^/]+/)?(?<modName>[^/]+$)))")
```
Comment (weshaggard) [inline eng/scripts/Language-Settings.ps1:40]: Why this change? We intentionally only had resourcemanager because they were the only approved projects that were deeper in paths.

File: eng/scripts/validate_go_mod.ps1 [hunk 3]
```diff
@@ -2,22 +2,29 @@ Param(
     [string] $serviceDir
 )
 
+. (Join-Path $PSScriptRoot . Language-Settings.ps1)
```
Comment (weshaggard) [inline eng/scripts/validate_go_mod.ps1:5]:
```suggestion
. (Join-Path $PSScriptRoot .. common scripts common.ps1)
```

File: eng/scripts/validate_go_mod.ps1 [hunk 4]
```diff
@@ -2,22 +2,29 @@ Param(
...
 
 $hasError = $false
 foreach ($goMod in $goModFiles) {
-    $name = $goMod.FullName
-    $patternMatches = Get-Content $name | Select-String -Pattern "replace "
-    if ($patternMatches.Length -ne 0) {
-        Write-Host "Found a replace directive in go.mod file at $name"
-        $hasError = $true
+    $mod = Get-GoModuleProperties ($goMod -replace ".*\/azure-sdk-for-go\/")
+    if ($mod -ne $null) {
```
Comment (weshaggard) [inline eng/scripts/validate_go_mod.ps1:19]:
```suggestion
    if ($mod) {
```

File: eng/scripts/Language-Settings.ps1 (hunk 2 above)
Comment (seankane-msft) [reply eng/scripts/Language-Settings.ps1:40]: @benbp helped out here. This regex does not match for the multi directory (ie sdk/data/aztables). It will match for sdk/azcore and sdk/azidentity, but not further. Is there a need to keep the resourcemanager portion?

Comment (weshaggard) [reply eng/scripts/Language-Settings.ps1:40]: This should be matching sdk/data/aztables just fine, if it weren't then that pipeline would be completely broken. Is there some evidence that it isn't working for them?

Comment (seankane-msft) [reply eng/scripts/Language-Settings.ps1:40]: @weshaggard https://regex101.com/r/yE8yp3/1

Comment (benbp) [reply eng/scripts/Language-Settings.ps1:40]: @weshaggard I agree that it's worth investigating why the pipeline is working because this regex does not appear to work based on our testing in that regex101 link.

Comment (weshaggard) [reply eng/scripts/Language-Settings.ps1:40]: Generally the path passed here is a directory not the full path to go.mod, see https://github.com/Azure/azure-sdk-for-go/blob/main/eng/scripts/Language-Settings.ps1#L108 as an example. If we want to support the full path as well we could add `(/go.mod)?` at the end of the regex.

Comment (seankane-msft) [reply eng/scripts/Language-Settings.ps1]: I passed in the directory and this worked as expected. Thanks for clarifying @weshaggard

**Review by weshaggard — APPROVED** [review]
